src/macaron/repo_finder/repo_finder.py

```diff
@@ -0,0 +1,182 @@
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
+
+"""
+This module contains the logic for using/calling the different repo finders.
+
+Input
+-----
+The entry point of the repo finder depends on the type of PURL being analyzed.
+- If passing a PURL representing an artifact, the ``find_repo`` function in this file should be called.
+- If passing a PURL representing a repository, the ``to_repo_path`` function in this file should be called.
+
+Artifact PURLs
+--------------
+For artifact PURLs, the PURL type determines how the repositories are searched for.
+Currently, for Maven PURLs, SCM meta data is retrieved from the matching POM retrieved from Maven Central (or
+other configured location).
+
+For Python, .NET, Rust, and NodeJS type PURLs, Google's Open Source Insights API is used to find the meta data.
+
+In either case, any repository links are extracted from the meta data, then checked for validity via
+``DependencyAnalyzer::find_valid_url`` which accepts URLs that point to a Github repository or similar.
+
+Repository PURLs
+----------------
+For repository PURLs, the type is checked against the configured valid domains, and accepted or rejected based
+on that data.
+
+Result
+------
+If all goes well, a repository URL that matches the initial artifact or repository PURL will be returned for
+analysis.
+"""
+
+import logging
+import os
+from collections.abc import Iterable
+from urllib.parse import ParseResult, urlunparse
+
+from packageurl import PackageURL
+
+from macaron.config.defaults import defaults
+from macaron.repo_finder.repo_finder_base import BaseRepoFinder
+from macaron.repo_finder.repo_finder_deps_dev import DepsDevRepoFinder
+from macaron.repo_finder.repo_finder_java import JavaRepoFinder
+from macaron.slsa_analyzer.git_url import get_remote_vcs_url
+
+logger: logging.Logger = logging.getLogger(__name__)
+
+
+def find_repo(purl: PackageURL) -> str:
+    """Retrieve the repository URL that matches the given PURL.
+
+    Parameters
+    ----------
+    purl : PackageURL
+        The parsed PURL to convert to the repository path.
+
+    Returns
+    -------
+    str :
+        The repository URL found for the passed package.
+    """
+    repo_finder: BaseRepoFinder
+    if purl.type == "maven":
+        repo_finder = JavaRepoFinder()
+    elif defaults.getboolean("repofinder", "use_open_source_insights") and purl.type in [
+        "pypi",
+        "nuget",
+        "cargo",
+        "npm",
+    ]:
+        repo_finder = DepsDevRepoFinder()
+    else:
+        logger.debug("No Repo Finder found for package type: %s of %s", purl.type, purl.to_string())
+        return ""
+
+    # Call Repo Finder and return first valid URL
+    logger.debug("Analyzing %s with Repo Finder: %s", purl.to_string(), repo_finder.__class__)
+    for urls in repo_finder.find_repo(purl):
+        url = find_valid_url(urls)
+        if url:
+            return url
+
+    return ""
```

Comment:
Related to [this comment](https://github.com/oracle/macaron/pull/388/files#r1316858675) about the return type of the `BaseRepoFinder::find_repo` method: if the return type is just `Iterator[str]`, then I think here you can simplify this quite a bit:
```suggestion
    return find_valid_url(repo_finder.find_repo(purl))
```